packages/flutter_goldens/lib/src/flutter_goldens_io.dart

```diff
@@ -20,28 +22,110 @@ export 'package:flutter_goldens_client/skia_client.dart';
+  } else if (key is String) {
+    return MatchesGoldenFile.forStringPath(key, version, comparator: isFlaky ? _flakyGoldenFileComparator : null);
+  }
+  throw ArgumentError('Unexpected type for golden file: ${key.runtimeType}');
+}
+
+String? _namePrefix;
+late GoldenComparatorType _flakyGoldenFileComparator;
+
 /// Main method that can be used in a `flutter_test_config.dart` file to set
 /// [goldenFileComparator] to an instance of [FlutterGoldenFileComparator] that
 /// works for the current test. _Which_ FlutterGoldenFileComparator is
 /// instantiated is based on the current testing environment.
 ///
 /// When set, the `namePrefix` is prepended to the names of all gold images.
 Future<void> testExecutable(FutureOr<void> Function() testMain, {String? namePrefix}) async {
+  _namePrefix = namePrefix;
+  final GoldenFileComparator nonFlakyComparator = await _createComparator();
+  _flakyGoldenFileComparator = await _createComparator(isFlaky: true);
+
+  // Assign this last because FlutterLocalFileComparator.fromDefaultComparator
+  // is called twice, and it needs the original value of [goldenFileComparator].
+  goldenFileComparator = nonFlakyComparator;
+  await testMain();
+}
+
+/// Processes golden check commands sent from the browser process.
+///
+/// When running browser tests, goldens are not generated within the app itself
+/// due to browser restrictions. Instead, when a test calls [matchesFlutterGolden]
+/// the browser sends a [command] to a host process. This function handles the
+/// command.
+///
+/// This custom command handler is used for Flutter's own goldens. It
+/// understands the "isFlaky" property, a boolean encoded in the command as a
+/// custom command property. If true, uses a golden comparator that submits the
+/// golden, but does not fail the test, allowing manual inspection in the Skia
+/// Gold UI and verification of fixes.
+///
+/// See also:
+///   * [FlakyWebGoldenComparator], which implements custom browser-side logic.
+Future<void> processBrowserCommand(dynamic command) async {
+  if (command is Map<String, dynamic>) {
+    final io.File imageFile = io.File(command['imageFile'] as String);
+    final Uri goldenKey = Uri.parse(command['key'] as String);
+    final bool update = command['update'] as bool;
+    final Map<String, dynamic> customProperties = (command['customProperties'] as Map<String, dynamic>?) ?? const <String, dynamic>{};
+    final bool isFlaky = (customProperties['isFlaky'] as bool?) ?? false;
+    final Uint8List bytes = await io.File(imageFile.path).readAsBytes();
+    if (update) {
+      await goldenFileComparator.update(goldenKey, bytes);
+      io.stdout.writeln(jsonEncode(<String, dynamic>{'success': true}));
+    } else {
+      try {
+        final bool success = isFlaky
+          ? await _flakyGoldenFileComparator.compare(bytes, goldenKey)
+          : await goldenFileComparator.compare(bytes, goldenKey);
```

Comment:
Yeah, like right here.. instead of having separate comparators, why not just add `compareFlaky` as a method to the base abstract FlutterGoldenFileComparator class from flutter_goldens? The post-, pre-submit, and local comparators will be able to override it if they need to do anything special in each test environment.

It looks like it could eliminate a bunch of code creating and managing multiple comparators.

Reply:
I just realized that `compare` is also [called by the matcher](https://github.com/flutter/flutter/blob/23401885bf3e5abf102d7930fafb8d5a0c96a7cf/packages/flutter_test/lib/src/_matchers_io.dart#L73), which knows nothing about flakiness. So even if I have `compareFlaky` or an override of `compare` with `isFlaky` parameter, the matcher will not specify it 🤔

Reply:
Aren't you creating a custom matcher though? I may not be understanding this properly.

Reply:
I'm still reusing the existing matchers. I just teach them to [optionally take a custom comparator](https://github.com/flutter/flutter/pull/113396/files#diff-503cfe6d62ac4cd45c0b021f709e278bc212e3e185027e50dbecae7a4bc3417dR53) and use it instead of the default one.

I could make a custom copy of the matcher too. But now I'm starting to realize that by replacing `matchesGoldenFile`, the comparator, and the matcher, I'd pretty much be duplicating all existing functionality. Is that OK? Or should we look for a way to reuse and customize existing functionality? An idea:

* Go back to https://github.com/flutter/flutter/pull/111595, but instead of hard-coding `isFlaky`, we expose an optional named `String? tag` argument in `matchesGoldenFile`.
* This argument would be passed to `GoldenFileComparator` unchanged.
* We document that this argument is not used by the default implementation of the golden API but may be used in conjunction with `flutter_test_config.dart` to customize the functionality of custom comparators.

In the Flutter repo `tag: 'flaky'` would be implemented by the custom comparator to implement flaky goldens, e.g.:

```dart
matchesGoldenFile('date_picker_test.time.initial.png', tag: 'flaky')
```

Properties of this approach:

* Pro: it does not hard-code the notion of "flaky" into the public golden API.
* Pro: it reuses all of the existing golden functionality, so that customizations like this are easy to implement.
* Pro: it's flexible enough for developer to customize for different things (it's just a string; you can put whatever you want in it).
* Con: `String`-based API is not very strongly typed (can be alleviated by checking the value at runtime).
* Con: it's breaking, but easily fixable (simply pass `customProperties` to `flutter_test_config.dart`, or ignore it if it's not needed).

Reply:
> But now I'm starting to realize that by replacing matchesGoldenFile, the comparator, and the matcher, I'd pretty much be duplicating all existing functionality.

I don't think you need to replace all of these things. We already have custom comparators. They do not need to be copied and rewritten. If we create a `matchesFlutterGolden` method for the specific purpose of the flutter framework, then it can call on the custom comparator (of type FlutterGoldenFileComparator) that we already have to execute the test however we want it to. Because it is already encapsulated for our own specific uses. I feel like a lot of this is not necessary.

Reply:
See https://github.com/flutter/flutter/pull/114026 as a rough example of what I mean.

Reply:
Ah, you're right, I forgot that custom comparators can be reused, so we're not duplicating everything. Ok, moving in that direction.